File: AdventOfCode2021/pysolns/d11.py
import sys
from collections import defaultdict
from itertools import product

mvs = [pt for pt in product(range(-1,2),range(-1,2)) if pt != (0,0)]

def lines():
  return [line.strip() for line in sys.stdin]

obrd = [[int(x) for x in l] for l in lines()]
H = len(obrd)
W = len(obrd[0])
pts = list(product(range(H), range(W)))
obrd = {(x,y): obrd[x][y] for (x,y) in pts}

def valid(pt):
  return 0 <= pt[0] < H and 0 <= pt[1] < W

def ns(pt):
  return [p for p in [(pt[0]+x,pt[1]+y) for (x,y) in mvs] if valid(p)]

def flashprop(b):
  flashed = set()
  toproc = [pt for pt in pts if b[pt] > 9]
  cnt = 0
  while toproc:
    pt = toproc.pop()
    if pt in flashed:
      continue
    cnt += 1
    flashed.add(pt)
    for n in ns(pt):
      b[n] += 1
      if b[n] > 9 and n not in flashed:
        toproc.append(n)
  return cnt

def procstep(b):
  for pt in pts:
    b[pt] += 1
  cnt = flashprop(b)
  for pt in pts:
    if b[pt] > 9:
      b[pt] = 0
  return cnt

def partA():
  brd = {k:v for k,v in obrd.items()}
  sm = 0
  for _ in range(100):
    sm += procstep(brd)
  print(sm)

def partB():
  brd = {k:v for k,v in obrd.items()}
  stp = 0
  while True:
    stp += 1
    procstep(brd)
    if all(v == 0 for v in brd.values()):
      break
  print(stp)

partA()
partB()

File: AdventOfCode2021/d11_julia/d11.jl
mvs = [[0, 1], [0, -1], [1, 0], [-1, 0], [1, 1], [1, -1], [-1, 1], [-1, -1]]
function readinp()
    inp = []
    while true
        l = readline()
        if l == ""
            break
        end
        push!(inp, [parse(Int, x) for x=chomp(l)])
    end
    return hcat(inp...)'
end

obrd = readinp()
H = size(obrd, 1)
W = size(obrd, 1)

function valid(pt)
    return 1 <= pt[1] <= H && 1 <= pt[2] <= W
end

function ns(pt)
    [pt + x for x=mvs if valid(pt + x)]
end

function flashprop(board)
    q = [[x,y] for x=1:H for y=1:W if board[x,y] > 9]
    s = Set()
    cnt = 0
    while length(q) > 0
        hd = pop!(q)
        if in(hd, s)
            continue
        end
        cnt += 1
        push!(s, hd)
        for n=ns(hd)
            board[n[1],n[2]] += 1
            if board[n[1], n[2]] > 9 && !in(n, s)
                push!(q, n)
            end
        end
    end
    return cnt
end

function procstep(board)
    for i=1:H
        for j=1:W
            board[i,j] += 1
        end
    end
    cnt = flashprop(board)
    for i=1:H
        for j=1:W
            if board[i,j] > 9
                board[i, j] = 0
            end
        end
    end
    return cnt
end

function part1()
    b = copy(obrd)
    sm = 0
    for i=1:100
        sm += procstep(b)
    end
    println(sm)
end

function part2()
    b = copy(obrd)
    stp = 0
    while true
        stp += 1
        procstep(b)
        all = true
        for i=1:H
            for j=1:W
                if b[i,j] != 0
                    all = false
                end
            end
        end
        if all
            break
        end
    end
    println(stp)
end

part1()
part2()
